### ISSUE TITLE
NaN Output in Distributed Model with Quantization and Weighted EmbeddingBagCollection Layer

### ISSUE THREAD

**Issue by jiannanWang** [meta-pytorch/torchrec#1130]
### Content
I have been training a quantized model with weighted `EmbeddingBagCollection` in the nightly version of pytorch (2.1.0.dev20230423) and torchrec-nightly (2023.4.23) and have encountered an issue where the model produces NaN results after the first epoch .

Upon further investigation, I discovered that during the forward pass, the model's `EmbeddingBagCollection` layer produces some NaN in its output, which in turn causes the subsequent layer to produce all NaN output. This issue seems to occur only when the model is distributed across multiple GPUs, as I did not encounter any NaN output when running the model and the same input on a single GPU.

I have included a reproduction code with world_size set to 4 to demonstrate the issue. I made a model with a dense layer, an `EmbeddingBagCollection` layer, a weighted `EmbeddingBagCollection` layer, and an over layer. The code first quantized the model, then distributed the model, copy the model’s state_dict to make sure the distributed model’s weight is the same as the original model, and finally run one forward pass of the distributed model and the non-distributed model and compare the model’s results.  When running this code in the nightly version of PyTorch and TorchRec, the output shows that the sparse weighted layer (weighted `EmbeddingBagCollection` layer) contains NaN, and subsequently, the over layer and prediction results become all NaN in ranks 1, 2, and 3. While for the nondistributed model, there’s no NaN in its results.

I believe this issue may be a bug, as I would expect the behavior of the distributed model to be the same as that of the non-distributed model (no NaN in the results).

### Reproduction code
```
import os
import copy
import random
from typing import Callable, List, Tuple, Union
import numpy as np
import multiprocessing

import torch
import torch.distributed as dist
import torch.nn as nn
from torchrec.modules.embedding_configs import EmbeddingBagConfig
from torchrec.modules.embedding_modules import EmbeddingBagCollection
from torchrec.distributed.model_parallel import DistributedModelParallel
from torchrec.distributed.planner import EmbeddingShardingPlanner, Topology
from torchrec.distributed.quant_embeddingbag import QuantEmbeddingBagCollectionSharder
from torchrec.distributed.test_utils.multi_process import MultiProcessContext
from torchrec.distributed.test_utils.test_sharding import copy_state_dict
from torchrec.distributed.test_utils.test_model import ModelInput
from torchrec.distributed.types import ModuleSharder, ShardingEnv, ShardingPlan
from torchrec.inference.modules import quantize_embeddings
from torchrec.sparse.jagged_tensor import KeyedTensor
from torchrec.test_utils import get_free_port

def setUp():
    os.environ["MASTER_ADDR"] = str("localhost")
    os.environ["MASTER_PORT"] = str(get_free_port())
    os.environ["GLOO_DEVICE_TRANSPORT"] = "TCP"
    os.environ["NCCL_SOCKET_IFNAME"] = "lo"

    torch.use_deterministic_algorithms(True)
    if torch.cuda.is_available():
        torch.backends.cudnn.allow_tf32 = False
        torch.backends.cuda.matmul.allow_tf32 = False
        os.environ["CUBLAS_WORKSPACE_CONFIG"] = ":4096:8"


class TestModel(nn.Module):
    def __init__(self):
        super().__init__()
        # define model parameters
        self.dense_in_feature = 311
        self.dense_out_feature = 478
        self.table_params = [
            [631, 720],
        ]
        self.weighted_table_params = [
            [278, 756],
            [262, 300],
            [752, 76],
            [675, 464],
        ]
        self.over_out_feature = 701

        # sparse layer
        self.tables = [
            EmbeddingBagConfig(
                num_embeddings=self.table_params[i][0],
                embedding_dim=self.table_params[i][1],
                name="table_" + str(i),
                feature_names=["feature_" + str(i)],
            )
            for i in range(len(self.table_params))
        ]
        self.sparse = EmbeddingBagCollection(
            tables=self.tables,
            is_weighted=False,
        )
        # weighted sparse layer
        self.weighted_tables = [
            EmbeddingBagConfig(
                num_embeddings=self.weighted_table_params[i][0],
                embedding_dim=self.weighted_table_params[i][1],
                name="weighted_table_" + str(i),
                feature_names=["weighted_feature_" + str(i)],
            )
            for i in range(len(self.weighted_table_params))
        ]
        self.sparse_weighted = EmbeddingBagCollection(
            tables=self.weighted_tables, 
            is_weighted=True,
        )
        # dense layer
        self.dense = nn.Linear(in_features=self.dense_in_feature, out_features=self.dense_out_feature, bias=True)
        # over layer
        in_features_concat = (
            self.dense_out_feature
            + sum([table.embedding_dim * len(table.feature_names) for table in self.tables])
            + sum([table.embedding_dim * len(table.feature_names) for table in self.weighted_tables])
        )
        self.over = nn.Linear(in_features=in_features_concat, out_features=self.over_out_feature, bias=True)

    def forward(
        self,
        input: ModelInput,
    ) -> Union[torch.Tensor, Tuple[torch.Tensor, torch.Tensor]]:
        # dense, sparse, weighted sparse layer output
        dense_r = self.dense(input.float_features)
        sparse_r = self.sparse(input.idlist_features)
        sparse_weighted_r = self.sparse_weighted(input.idscore_features)
        # concat dense, sparse, weighted sparse layer output
        result = KeyedTensor(
            keys=sparse_r.keys() + sparse_weighted_r.keys(),
            length_per_key=sparse_r.length_per_key()
            + sparse_weighted_r.length_per_key(),
            values=torch.cat([sparse_r.values(), sparse_weighted_r.values()], dim=1),
        )
        _features = [feature for table in self.tables for feature in table.feature_names]
        _weighted_features = [feature for table in self.weighted_tables for feature in table.feature_names]

        ret_list = []
        ret_list.append(dense_r)
        for feature_name in _features:
            ret_list.append(result[feature_name])
        for feature_name in _weighted_features:
            ret_list.append(result[feature_name])
        ret_concat = torch.cat(ret_list, dim=1)
        # over layer output
        over_r = self.over(ret_concat)
        # sigmoid output
        pred = torch.sigmoid(torch.mean(over_r, dim=1))
        # return model's output and layers' output
        return pred, (dense_r, sparse_r, sparse_weighted_r, over_r)

def count_nan(tensor):
    # count nan in tensor
    return torch.sum(torch.isnan(tensor)).item()

def sharding_single_rank_test(
    rank: int,
    world_size: int,
    model,
    inputs,
    sharders: List[ModuleSharder[nn.Module]],
    backend: str,
    num_iters = 1,
) -> None:
    with MultiProcessContext(rank, world_size, backend) as ctx:
        # copy model to ctx.device and quantize embedding
        model = model.to(ctx.device)
        model = quantize_embeddings(model, dtype=torch.qint8, inplace=True)
        
        # copy model to global_model
        global_model = copy.deepcopy(model)
        global_input = inputs[0][0].to(ctx.device)

        # copy model to local_model and sharding
        local_model = copy.deepcopy(model)
            
        planner = EmbeddingShardingPlanner(topology=Topology(world_size, ctx.device.type))
        plan: ShardingPlan = planner.collective_plan(local_model, sharders, ctx.pg)

        local_model = DistributedModelParallel(
            local_model,
            env=ShardingEnv.from_process_group(ctx.pg),
            plan=plan,
            sharders=sharders,
            device=ctx.device,
        )

        # copy global_model's state_dict to local_model
        copy_state_dict(local_model.state_dict(), model.state_dict())

        local_input = inputs[0][1][rank].to(ctx.device)
        local_model.eval()

        # run a single forward step of the sharded model
        for _ in range(num_iters):
            local_pred, (dense_r, sparse_r, sparse_weighted_r, over_r) = local_model(local_input)

        # collect all local_pred, dense_r, sparse_r, sparse_weighted_r, over_r from all ranks
        all_local_pred = []
        for _ in range(world_size):
            all_local_pred.append(torch.empty_like(local_pred))
        dist.all_gather(all_local_pred, local_pred, group=ctx.pg)
        
        all_dense_r = []
        for _ in range(world_size):
            all_dense_r.append(torch.empty_like(dense_r))
        dist.all_gather(all_dense_r, dense_r, group=ctx.pg)

        sparse_r_dict = sparse_r.to_dict()
        all_sparse_r_dict = {}
        for key in sparse_r_dict:
            all_sparse_r_dict[key] = []
            for _ in range(world_size):
                all_sparse_r_dict[key].append(torch.empty_like(sparse_r_dict[key]))
            dist.all_gather(all_sparse_r_dict[key], sparse_r_dict[key].contiguous(), group=ctx.pg)

        sparse_weighted_r_dict = sparse_weighted_r.to_dict()
        all_sparse_weighted_r_dict = {}
        for key in sparse_weighted_r_dict:
            all_sparse_weighted_r_dict[key] = []
            for _ in range(world_size):
                all_sparse_weighted_r_dict[key].append(torch.empty_like(sparse_weighted_r_dict[key]))
            dist.all_gather(all_sparse_weighted_r_dict[key], sparse_weighted_r_dict[key].contiguous(), group=ctx.pg)

        all_over_r = []
        for _ in range(world_size):
            all_over_r.append(torch.empty_like(over_r))
        dist.all_gather(all_over_r, over_r, group=ctx.pg)

        # run a single forward step of the nondistributed model
        global_model.eval()
        for _ in range(num_iters):
            global_pred, (global_dense_r, global_sparse_r, global_sparse_weighted_r, global_over_r) = global_model(global_input)

        if rank == 0:
            # check nan in all layers output
            print("NaN check for distributed model")
            for r in range(world_size):
                print("\tcount nan in all dense r in rank ", r, " : ",  count_nan(all_dense_r[r]), "/", all_dense_r[r].numel())
            for r in range(world_size):
                sum_nan = 0
                sum_element = 0
                for key in all_sparse_r_dict:
                    temp_nan = count_nan(all_sparse_r_dict[key][r])
                    sum_nan += temp_nan
                    temp_element = all_sparse_r_dict[key][r].numel()
                    sum_element += temp_element
                print("\tcount nan in all sparse r in rank ", r, " : ",  sum_nan, "/", sum_element)
            for r in range(world_size):
                sum_nan = 0
                sum_element = 0
                for key in all_sparse_weighted_r_dict:
                    temp_nan = count_nan(all_sparse_weighted_r_dict[key][r])
                    sum_nan += temp_nan
                    temp_element = all_sparse_weighted_r_dict[key][r].numel()
                    sum_element += temp_element
                print("\tcount nan in all sparse weighted r in rank ", r, " : ",  sum_nan, "/", sum_element)
            for r in range(world_size):
                print("\tcount nan in all over r in rank ", r, " : ",  count_nan(all_over_r[r]), "/", all_over_r[r].numel())
            for r in range(world_size):
                print("\tcount nan in all local pred in rank ", r, " : ",  count_nan(all_local_pred[r]), "/", all_local_pred[r].numel())

            print("NaN check for non distributed model")
            print("\tcount nan in dense r : ",  count_nan(global_dense_r), "/", global_dense_r.numel())
            global_sparse_r_dict = global_sparse_r.to_dict()
            sum_nan = 0
            sum_element = 0
            for key in global_sparse_r_dict:
                temp_nan = count_nan(global_sparse_r_dict[key])
                sum_nan += temp_nan
                temp_element = global_sparse_r_dict[key].numel()
                sum_element += temp_element
            print("\tcount nan in sparse r : ",  sum_nan, "/", sum_element)
            global_sparse_weighted_r_dict = global_sparse_weighted_r.to_dict()
            sum_nan = 0
            sum_element = 0
            for key in global_sparse_weighted_r_dict:
                temp_nan = count_nan(global_sparse_weighted_r_dict[key])
                sum_nan += temp_nan
                temp_element = global_sparse_weighted_r_dict[key].numel()
                sum_element += temp_element
            print("\tcount nan in sparse weighted r : ",  sum_nan, "/", sum_element)
            print("\tcount nan in over r : ",  count_nan(global_over_r), "/", global_over_r.numel())
            print("\tcount nan in global pred : ",  count_nan(global_pred), "/", global_pred.numel())


def run_multi_process_test(
    callable: Callable[
        ...,
        None,
    ],
    world_size: int,
    # pyre-ignore
    **kwargs,
) -> None:
    setUp()

    ctx = multiprocessing.get_context("forkserver")
    processes = []
    for rank in range(world_size):
        kwargs["rank"] = rank
        kwargs["world_size"] = world_size
        p = ctx.Process(
            target=callable,
            kwargs=kwargs,
        )
        p.start()
        processes.append(p)

    for p in processes:
        p.join()

def main_test(
    backend,
    world_size,
) -> None:
    # create model and input
    model = TestModel()
    inputs = [ModelInput.generate(
        batch_size=200,
        world_size=world_size,
        num_float_features=model.dense_in_feature,
        tables=model.tables,
        weighted_tables=model.weighted_tables,
    )]
    # create sharders
    sharders = [
            QuantEmbeddingBagCollectionSharder(),
        ]
    # run test
    run_multi_process_test(
        callable=sharding_single_rank_test,
        world_size=world_size,
        model=model,
        inputs=inputs,
        sharders=sharders,
        backend=backend,
    )


if __name__ == "__main__":
    # fix seed for reproducibility
    seed = 0
    torch.manual_seed(seed)
    random.seed(seed)
    np.random.seed(seed)
    # set nccl backend and world size
    main_test(backend="nccl", world_size=4)
```

### Output
```
NaN check for distributed model
        count nan in all dense r in rank  0  :  0 / 95600
        count nan in all dense r in rank  1  :  0 / 95600
        count nan in all dense r in rank  2  :  0 / 95600
        count nan in all dense r in rank  3  :  0 / 95600
        count nan in all sparse r in rank  0  :  0 / 144000
        count nan in all sparse r in rank  1  :  0 / 144000
        count nan in all sparse r in rank  2  :  0 / 144000
        count nan in all sparse r in rank  3  :  0 / 144000
        count nan in all sparse weighted r in rank  0  :  0 / 319200
        count nan in all sparse weighted r in rank  1  :  4714 / 319200
        count nan in all sparse weighted r in rank  2  :  2874 / 319200
        count nan in all sparse weighted r in rank  3  :  2249 / 319200
        count nan in all over r in rank  0  :  0 / 140200
        count nan in all over r in rank  1  :  140200 / 140200
        count nan in all over r in rank  2  :  140200 / 140200
        count nan in all over r in rank  3  :  140200 / 140200
        count nan in all local pred in rank  0  :  0 / 200
        count nan in all local pred in rank  1  :  200 / 200
        count nan in all local pred in rank  2  :  200 / 200
        count nan in all local pred in rank  3  :  200 / 200
NaN check for non distributed model
        count nan in dense r :  0 / 382400
        count nan in sparse r :  0 / 576000
        count nan in sparse weighted r :  0 / 1276800
        count nan in over r :  0 / 560800
        count nan in global pred :  0 / 800
```

**Comment by YLGH** [comment]
Hey Jiannan, I think this issue is related to https://github.com/pytorch/torchrec/issues/1131 - closing this one for now, let's continue discussion in other thread

**Comment by jiannanWang** [comment]
Can we reopen this issue? Because it still exists after applying the fix in #1131.

Hi, thank you for your fix in the other thread! However, it seems not to solve this NaN issue.

I updated my reproduction code below using `from torchrec.distributed.shard import _shard_modules` instead of `from torchrec.distributed.model_parallel import DistributedModelParallel`. The behavior is still the same: the model's `EmbeddingBagCollection` layer produces some NaN in its output, which in turn causes the subsequent layer to produce all NaN output. 

Below is the updated reproduction code and the program output:

### Reproduction code
```
import os
import copy
import random
from typing import Callable, List, Tuple, Union
import numpy as np
import multiprocessing

import torch
import torch.distributed as dist
import torch.nn as nn
from torchrec.modules.embedding_configs import EmbeddingBagConfig
from torchrec.modules.embedding_modules import EmbeddingBagCollection
from torchrec.distributed.model_parallel import DistributedModelParallel
from torchrec.distributed.planner import EmbeddingShardingPlanner, Topology
from torchrec.distributed.quant_embeddingbag import QuantEmbeddingBagCollectionSharder
from torchrec.distributed.test_utils.multi_process import MultiProcessContext
from torchrec.distributed.test_utils.test_sharding import copy_state_dict
from torchrec.distributed.test_utils.test_model import ModelInput
from torchrec.distributed.types import ModuleSharder, ShardingEnv, ShardingPlan
from torchrec.inference.modules import quantize_embeddings
from torchrec.sparse.jagged_tensor import KeyedTensor
from torchrec.test_utils import get_free_port

def setUp():
    os.environ["MASTER_ADDR"] = str("localhost")
    os.environ["MASTER_PORT"] = str(get_free_port())
    os.environ["GLOO_DEVICE_TRANSPORT"] = "TCP"
    os.environ["NCCL_SOCKET_IFNAME"] = "lo"

    torch.use_deterministic_algorithms(True)
    if torch.cuda.is_available():
        torch.backends.cudnn.allow_tf32 = False
        torch.backends.cuda.matmul.allow_tf32 = False
        os.environ["CUBLAS_WORKSPACE_CONFIG"] = ":4096:8"


class TestModel(nn.Module):
    def __init__(self):
        super().__init__()
        # define model parameters
        self.dense_in_feature = 311
        self.dense_out_feature = 478
        self.table_params = [
            [631, 720],
        ]
        self.weighted_table_params = [
            [278, 756],
            [262, 300],
            [752, 76],
            [675, 464],
        ]
        self.over_out_feature = 701

        # sparse layer
        self.tables = [
            EmbeddingBagConfig(
                num_embeddings=self.table_params[i][0],
                embedding_dim=self.table_params[i][1],
                name="table_" + str(i),
                feature_names=["feature_" + str(i)],
            )
            for i in range(len(self.table_params))
        ]
        self.sparse = EmbeddingBagCollection(
            tables=self.tables,
            is_weighted=False,
        )
        # weighted sparse layer
        self.weighted_tables = [
            EmbeddingBagConfig(
                num_embeddings=self.weighted_table_params[i][0],
                embedding_dim=self.weighted_table_params[i][1],
                name="weighted_table_" + str(i),
                feature_names=["weighted_feature_" + str(i)],
            )
            for i in range(len(self.weighted_table_params))
        ]
        self.sparse_weighted = EmbeddingBagCollection(
            tables=self.weighted_tables, 
            is_weighted=True,
        )
        # dense layer
        self.dense = nn.Linear(in_features=self.dense_in_feature, out_features=self.dense_out_feature, bias=True)
        # over layer
        in_features_concat = (
            self.dense_out_feature
            + sum([table.embedding_dim * len(table.feature_names) for table in self.tables])
            + sum([table.embedding_dim * len(table.feature_names) for table in self.weighted_tables])
        )
        self.over = nn.Linear(in_features=in_features_concat, out_features=self.over_out_feature, bias=True)

    def forward(
        self,
        input: ModelInput,
    ) -> Union[torch.Tensor, Tuple[torch.Tensor, torch.Tensor]]:
        # dense, sparse, weighted sparse layer output
        dense_r = self.dense(input.float_features)
        sparse_r = self.sparse(input.idlist_features)
        sparse_weighted_r = self.sparse_weighted(input.idscore_features)
        # concat dense, sparse, weighted sparse layer output
        result = KeyedTensor(
            keys=sparse_r.keys() + sparse_weighted_r.keys(),
            length_per_key=sparse_r.length_per_key()
            + sparse_weighted_r.length_per_key(),
            values=torch.cat([sparse_r.values(), sparse_weighted_r.values()], dim=1),
        )
        _features = [feature for table in self.tables for feature in table.feature_names]
        _weighted_features = [feature for table in self.weighted_tables for feature in table.feature_names]

        ret_list = []
        ret_list.append(dense_r)
        for feature_name in _features:
            ret_list.append(result[feature_name])
        for feature_name in _weighted_features:
            ret_list.append(result[feature_name])
        ret_concat = torch.cat(ret_list, dim=1)
        # over layer output
        over_r = self.over(ret_concat)
        # sigmoid output
        pred = torch.sigmoid(torch.mean(over_r, dim=1))
        # return model's output and layers' output
        return pred, (dense_r, sparse_r, sparse_weighted_r, over_r)

def count_nan(tensor):
    # count nan in tensor
    return torch.sum(torch.isnan(tensor)).item()

def sharding_single_rank_test(
    world_size: int,
    model,
    inputs,
    sharders: List[ModuleSharder[nn.Module]],
) -> None:
    model = model.to("cuda")
    model = quantize_embeddings(model, dtype=torch.qint8, inplace=True)
    device = torch.device("cuda:0")
    
    # copy model to global_model
    global_model = copy.deepcopy(model)
    global_input = inputs[0][0].to(device)

    # copy model to local_model and sharding
    local_model = copy.deepcopy(model)
        
    planner = EmbeddingShardingPlanner(topology=Topology(world_size, device.type))
    plan: ShardingPlan = planner.plan(local_model, sharders)

    from torchrec.distributed.shard import _shard_modules
    local_model = _shard_modules(
        local_model,
        env=ShardingEnv.from_local(world_size=world_size, rank=0),
        plan=plan,
        sharders=sharders,
        device=device,
    )

    # copy global_model's state_dict to local_model
    copy_state_dict(local_model.state_dict(), model.state_dict())

    local_model.eval()
    local_pred, (local_dense_r, local_sparse_r, local_sparse_weighted_r, local_over_r) = local_model(global_input)

    # run a single forward step of the nondistributed model
    global_model.eval()
    global_pred, (global_dense_r, global_sparse_r, global_sparse_weighted_r, global_over_r) = global_model(global_input)

    # check nan in all layers output
    print("NaN check for distributed model")
    print("\tcount nan in dense r : ",  count_nan(local_dense_r), "/", local_dense_r.numel())
    local_sparse_r_dict = local_sparse_r.to_dict()
    sum_nan = 0
    sum_element = 0
    for key in local_sparse_r_dict:
        temp_nan = count_nan(local_sparse_r_dict[key])
        sum_nan += temp_nan
        temp_element = local_sparse_r_dict[key].numel()
        sum_element += temp_element
    print("\tcount nan in sparse r : ",  sum_nan, "/", sum_element)
    local_sparse_weighted_r_dict = local_sparse_weighted_r.to_dict()
    sum_nan = 0
    sum_element = 0
    for key in local_sparse_weighted_r_dict:
        temp_nan = count_nan(local_sparse_weighted_r_dict[key])
        sum_nan += temp_nan
        temp_element = local_sparse_weighted_r_dict[key].numel()
        sum_element += temp_element
    print("\tcount nan in sparse weighted r : ",  sum_nan, "/", sum_element)
    print("\tcount nan in over r : ",  count_nan(local_over_r), "/", local_over_r.numel())
    print("\tcount nan in global pred : ",  count_nan(local_pred), "/", local_pred.numel())
    print("\tcount nan in local pred : ",  count_nan(local_pred), "/", local_pred.numel())

    print("NaN check for non distributed model")
    print("\tcount nan in dense r : ",  count_nan(global_dense_r), "/", global_dense_r.numel())
    global_sparse_r_dict = global_sparse_r.to_dict()
    sum_nan = 0
    sum_element = 0
    for key in global_sparse_r_dict:
        temp_nan = count_nan(global_sparse_r_dict[key])
        sum_nan += temp_nan
        temp_element = global_sparse_r_dict[key].numel()
        sum_element += temp_element
    print("\tcount nan in sparse r : ",  sum_nan, "/", sum_element)
    global_sparse_weighted_r_dict = global_sparse_weighted_r.to_dict()
    sum_nan = 0
    sum_element = 0
    for key in global_sparse_weighted_r_dict:
        temp_nan = count_nan(global_sparse_weighted_r_dict[key])
        sum_nan += temp_nan
        temp_element = global_sparse_weighted_r_dict[key].numel()
        sum_element += temp_element
    print("\tcount nan in sparse weighted r : ",  sum_nan, "/", sum_element)
    print("\tcount nan in over r : ",  count_nan(global_over_r), "/", global_over_r.numel())
    print("\tcount nan in global pred : ",  count_nan(global_pred), "/", global_pred.numel())


def main_test(
    world_size,
) -> None:
    # create model and input
    model = TestModel()
    inputs = [ModelInput.generate(
        batch_size=200,
        world_size=world_size,
        num_float_features=model.dense_in_feature,
        tables=model.tables,
        weighted_tables=model.weighted_tables,
    )]
    # create sharders
    sharders = [
            QuantEmbeddingBagCollectionSharder(),
        ]
    # run test
    sharding_single_rank_test(
        world_size=world_size,
        model=model,
        inputs=inputs,
        sharders=sharders,
    )


if __name__ == "__main__":
    # fix seed for reproducibility
    seed = 0
    torch.manual_seed(seed)
    random.seed(seed)
    np.random.seed(seed)
    # set nccl backend and world size
    main_test(world_size=4)
```

### Output
```
NaN check for distributed model
        count nan in dense r :  0 / 382400
        count nan in sparse r :  0 / 576000
        count nan in sparse weighted r :  20052 / 1276800
        count nan in over r :  560800 / 560800
        count nan in global pred :  800 / 800
        count nan in local pred :  800 / 800
NaN check for non distributed model
        count nan in dense r :  0 / 382400
        count nan in sparse r :  0 / 576000
        count nan in sparse weighted r :  0 / 1276800
        count nan in over r :  0 / 560800
        count nan in global pred :  0 / 800
```
